lib/spack/spack/build_systems/compiler.py

```diff
@@ -0,0 +1,154 @@
+import os
+import re
+import sys
+from typing import Sequence
+
+import llnl.util.tty as tty
+from llnl.util.lang import classproperty
+
+import spack.compiler
+import spack.package_base
+
+
+def true():
+    return True
+
+
+class CompilerPackage(spack.package_base.PackageBase):
+    """A Package mixin for all common logic for packages that implement compilers"""
+
+    # metadata identifying this as a compiler for lmod
+    family = "compiler"
+
+    # TODO: how do these play nicely with other tags
+    tags = ["compiler"]
+
+    # Optional suffix regexes for searching for this type of compiler.
+    # Suffixes are used by some frameworks, e.g. macports uses an '-mp-X.Y'
+    # version suffix for gcc.
+    suffixes: Sequence[str] = [r"-.*"]
+
+    # Optional prefix regexes for searching for this compiler
+    prefixes: Sequence[str] = []
+
+    #: Compiler argument(s) that produces version information
+    #: If multiple arguments, the earlier arguments must produce errors when invalid
+    #: Must be a hashable type (e.g. tuple, rather than list)
+    version_argument = "-dumpversion"
+
+    #: Return values to ignore when invoking the compiler to get its version
+    ignore_version_errors: Sequence[int] = ()
+
+    #: Regex used to extract version from compiler's output
+    version_regex = "(.*)"
+
+    #: Platform matcher for Platform objects supported by compiler
+    # The value must be callable
+    is_supported_on_platform = true
+
+    #: Static definition of languages supported by this class
+    compiler_languages = ["c", "cxx", "fortran"]
+
+    def __init__(self, spec):
+        super().__init__(spec)
+        assert set(self.supported_languages) <= set(self.compiler_languages)
+
+    #: Dynamic definition of languages supported by this package
+    @property
+    def supported_languages(self):
+        return self.compiler_languages
+
+    @classproperty
+    def compiler_names(cls):
+        """Construct list of compiler names from per-language names"""
+        names = []
+        for language in cls.compiler_languages:
+            names.extend(getattr(cls, f"{language}_names"))
+        return names
+
+    @classproperty
+    def executables(cls):
+        """Construct executables for external detection from names, prefixes, and suffixes."""
+        regexp_fmt = r"^({0}){1}({2})$"
+        prefixes = [""] + cls.prefixes
+        suffixes = [""] + cls.suffixes
+        if sys.platform == "win32":
+            ext = r"\.(?:exe|bat)"
+            suffixes += [suf + ext for suf in suffixes]
+        return [
+            regexp_fmt.format(prefix, re.escape(name), suffix)
+            for prefix, name, suffix in itertools.product(prefixes, cls.compiler_names, suffixes)
+        ]
+
+    @classmethod
+    def determine_version(cls, exe):
+        version_argument = cls.version_argument
+        if isinstance(version_argument, str):
+            version_argument = [version_argument]
+
+        for va in version_argument:
```

Comment:
We can then remove this:
```suggestion
        for va in cls.version_arguments:
```

Reply:
Let's also postpone the discussion on this. I still like a clear type better, but this is not a major issue.